File: 4_семестр/Вычислительные_Методы/Lab3_math/main.py
from math import cos, pi, sin, sqrt, pow, asinh


def f(x):
    return sin(x) * (1 + cos(x)) / sqrt(1 + pow(cos(x), 2))


def trap(func, a, b, k, x):
    h = 1.0 * (b - a) / pow(10, k)
    Ik = 0.5 * (func(a) + func(b))
    for i in range(1, int(pow(10, k))):
        Ik += func(x + i * h)

    return Ik * h


J = -1 + sqrt(2) + asinh(1)
print(J)
for i in range(1, 10):
    tmp = trap(f, 0, pi/2, i, 0)
    print('I(', i, ') = ', tmp)
    print('J - I(', i, ') = ', abs(J - tmp))


def d(x):
    return (-pow(sin(x), 2) + pow(cos(x), 4) +
     pow(cos(x), 3) + pow(cos(x), 2) +
     cos(x) * (1 + pow(sin(x), 2))) / (
        pow(1 + pow(cos(x), 2), 3 / 2))


D = d(0)
print('D = ', d(0))


def dif(func, a, b, k):
    h = 1.0 * (b - a) / pow(10, k)
    return (func(a) - func(a - h)) / h


for i in range(1, 15):
    tmp = dif(f, 0, pi/2, i)
    print('d(', i, ') = ', tmp)
    print('D - d(', i, ') = ', abs(D - tmp))


def half_trap(func, a, b, k, x):
    h = 1.0 * (b - a) / pow(10, k)
    xk = x + 0.5*h
    Ik = 0
    for i in range(1, int(pow(10, k))):
        Ik += func(x + i * h)
    return Ik*0.5*h


def simpson(func, a, b, k, eps, x):
    old_sum = trap(func, a, b, k, x)
    new_sum = half_trap(func, a, b, k, x)
    ans = (4 * new_sum - old_sum) / 3
    err_est = max(1, abs(ans))
    while (err_est > abs(eps * ans)):
        old_ans = ans
        old_sum = new_sum
        new_sum = half_trap(func, a, b, k, x)
        ans = (4 * new_sum - old_sum) / 3
        err_est = abs(old_ans - ans)
    return ans


for i in range(1, 10):
    tmp = simpson(f, 0, pi/2, i, 0.0001, 0)
    print('simp(', i, ') = ', tmp)
    print('J - simp(', i, ') = ', abs(J - tmp))

File: 4_семестр/Вычислительные_Методы/c++/Lab3_math/main.cpp
#include <iostream>
#include <cstdio>
#include <chrono>
#include <cmath>
#include <iomanip>


using namespace std;
using namespace chrono;

double f(double x)
{
    return sin(x) * (1 + cos(x)) / sqrt(1 + pow(cos(x), 2));
}



double trap(double & a, double & b, int & k, double & x)
{
    double h = 1.0 * (b - a) / pow(10, k);
    double Ik = 0.5 * (f(a) + f(b));
    for (int i = 1; i < int(pow(10, k)); i++)
    {
        Ik += f(x + i * h);
    }
    return Ik * h;
}


double d(double x)
{
    return (-pow(sin(x), 2) + pow(cos(x), 4) +
            pow(cos(x), 3) + pow(cos(x), 2) +
            cos(x) * (1 + pow(sin(x), 2))) / (
                   pow(1 + pow(cos(x), 2), 3.0 / 2.0));
}

double dif(double & a, double & b, int & k)
{
    double h = 1.0 * (b - a) / pow(10, k);
    return (f(a) - f(a - h)) / h;
}

double simpson_rule(double & a, double & b, int & k)
{
    double h = 1.0 * (b - a) / pow(10, k);
    double simp_sum = (f(a) + 4 * f(a + h) + f(b));
    for (int i = 1; i < int(pow(10, k) / 2.0); ++i) {
        simp_sum += 2.0 * f(a + (a + (2.0 * i) * h)) + 4.0 * f(a + (2.0 * i + 1.0) * h);
    }
    return simp_sum * h / 3.0;
}


int main() {
    chrono::steady_clock sc;
    double J = -1 + sqrt(2) + asinh(1);
    cout << setprecision(20) << "Absolute integral: " << J << endl;
    double a = 0;
    double b = M_PI_2;
    double x = 0;
//    for (int i = 1; i < 15; ++i) {
//        auto start = steady_clock::now();
//        double tmp = trap(a, b, i, x);
//        cout << "I(" << i << ") = " << tmp << endl;
//        cout << "J - I(" << i << ") = " << abs(J - tmp) << endl;
//        auto end = steady_clock::now();
//        auto time_span = static_cast<chrono::duration<double>>(end - start);   // measure time span between start & end
//        cout << "Operation took: " << time_span.count() << " seconds !!!" << endl;
//    }
//    cout << endl;
//
//    double D = d(0);
//    cout << setprecision(20) << "D = " << D << endl;
//    for (int i = 1; i < 15; ++i) {
//        auto start = steady_clock::now();
//        double tmp = dif(a, b, i);
//        cout << "d(" << i << ") = " << tmp << endl;
//        cout << "D - d(" << i << ") = " << abs(D - tmp) << endl;
//        auto end = steady_clock::now();
//        auto time_span = static_cast<chrono::duration<double>>(end - start);   // measure time span between start & end
//        cout << "Operation took: " << time_span.count() << " seconds !!!" << endl;
//    }

    cout << endl;
    for (int i = 1; i < 15; ++i) {
        auto start = steady_clock::now();
        double tmp = simpson_rule(a, b, i);
        cout << setprecision(20) << "S(" << i << ") = " << tmp << endl;
        cout << "J - S(" << i << ") = " << abs(J - tmp) << endl;
        auto end = steady_clock::now();
        auto time_span = static_cast<chrono::duration<double>>(end - start);   // measure time span between start & end
        cout << "Operation took: " << time_span.count() << " seconds !!!" << endl;
    }

    return 0;
}